### DISCUSSION TITLE
Possible to use Nitro without filesystem routing?

### DISCUSSION THREAD

**Discussion by chrber04** [nitrojs/nitro#2730] [Help Wanted]
Hi everyone, I was wondering if it's possible to use nitro without the file based routing? And if it's possible to create the server & boot up the application more programatically. 

My use case is that I want to use Nitro with Effect-TS. This is an example of how I got it working with h3:

```
import type { App } from "h3";
import { Context, Effect, Layer } from "effect";
import { createApp  } from "h3";
export class AppService extends Context.Tag("AppService")<AppService, App>() {
	static readonly Live = Layer.sync(AppService, () => createApp());
}
...
import { createServer } from "node:http";
import { Effect, Layer } from "effect";
import { toNodeListener } from "h3";

import { AppService } from "./app/app.service.js";
import { AppConfig } from "./lib/configs/app.config.js";

const ServerStart = Effect.gen(function* () {
	const app = yield* AppService;
	const appConfig = yield* AppConfig;

	yield* Effect.acquireRelease(
		Effect.sync(() => {
			const server = createServer(toNodeListener(app));
			server.listen(appConfig.port, () => {
				console.log(`Server listening on port ${appConfig.port}...`);
			});
			return server;
		}),
		(server) => {
			return Effect.sync(() => {
				console.log("Server closing...");
				server.close();
			});
		}
	);
});
export const ServerLive = Layer.scopedDiscard(ServerStart).pipe(Layer.provide(AppService.Live));
...
export const AppLive = Layer.provide(ServerLive, Layer.mergeAll(AppRouter, AppLayer));
Effect.runFork(Layer.launch(AppLive));
```

H3 is great but I can't help shake the feeling that Nitro is better for the long term since it's more feature-complete.

**Comment by pi0** [comment]
Surely you can directly use `nitro.config` to define your routes ([example](https://stackblitz.com/edit/unjs-nitro-starter-qxj3ae))

```ts
//https://nitro.unjs.io/config
export default defineNitroConfig({
  handlers: [
    {
      route: '/',
      handler: '~/my-routes/index.ts',
    },
  ],
});
```

**Reply by chrber04** [reply]
That's a good point and I do appreciate your super quick response, but I assume this means it isn't possible to programatically start up the app? 

In Effect it's crucial to provide all services through layers for proper scoping, resource management, and finalization on startup/shutdown. And all these layers must be provided to an Effect runtime: https://effect.website/docs/guides/runtime

**Reply by pi0** [reply]
I am not familiar with the effect project's philosophy. Nitro generates a standalone server output. You can use plugins to add startup logic and even use `node-handler` preset to consume handler output in a custom setup.

TBH i think if you really want to go with Effect, probably composing h3 and other unjs goodies might make better sense. The con is you loose nitro deployment presets and conventions but you gain full control over setup.

**Reply by chrber04** [reply]
That makes sense, at least now I know. I will continue with h3 and maybe I'll come out a better developer from being forced to get a deeper understanding - thank you so much for the response :)
